src/lib.rs

```diff
@@ -1,718 +1,923 @@
-                } else {
-                    5
-                }
-            }
-            Packet::StimulusPortPage(_) => 1,
-            Packet::EventCounter(_) => 2,
-            Packet::ExceptionTrace(_) => 3,
-            Packet::PeriodicPcSample(pps) => {
-                if pps.pc().is_some() {
-                    5
-                } else {
-                    2
-                }
-            }
-            Packet::DataTracePcValue(_) => 5,
-            Packet::DataTraceAddress(_) => 3,
-            Packet::DataTraceDataValue(dtdv) => 1 /* header */ + dtdv.size,
-        }
-    }
+    /// Next bytes will be assumed to be part of a GlobalTimestamp2
+    /// packet, until the MSB is set.
+    GlobalTimestamp2,
 }
 
-/// Tries to parse an ITM packet from the start of the given buffer
-fn parse(input: &[u8]) -> Result<Packet, Either<Error, NeedMoreBytes>> {
-    let header = input.first().cloned().ok_or(Either::Right(NeedMoreBytes))?;
+enum HeaderVariant {
+    Packet(TracePacket),
+    Stub(PacketStub),
+}
 
-    match Header::parse(header).map_err(Either::Left)? {
-        Header::Synchronization => {
-            let mut cursor = 1u8;
+pub struct DecoderOptions {
+    /// Whether to keep reading after a (temporary) EOF condition. If
+    /// set and `next` is called on a decoder iterator, `next` will
+    /// never return if unless the EOF condition is eventually resolved.
+    pub ignore_eof: bool,
+}
 
-            loop {
-                match input.get(usize::from(cursor)) {
-                    Some(&0b0000_0000) => {
-                        // still within the synchronization packet
-                        cursor += 1;
-                        continue;
-                    }
-                    Some(&0b1000_0000) if cursor >= 5 => {
-                        //  "Synchronization packet is at least forty-seven 0 bits followed by single 1
-                        //  bit"
-                        // valid synchronization packet
-                        break Ok(Packet::Synchronization(Synchronization { len: cursor + 1 }));
-                    }
-                    Some(_) => {
-                        // malformed packet
-                        break Err(Either::Left(Error::MalformedPacket {
-                            header,
-                            len: cursor,
-                        }));
-                    }
-                    None => {
-                        // need more bytes
-                        break Err(Either::Right(NeedMoreBytes));
-                    }
-                }
-            }
-        }
+impl Default for DecoderOptions {
+    fn default() -> Self {
+        Self { ignore_eof: true }
+    }
+}
 
-        // Overflow packets have no payload
-        Header::Overflow => Ok(Packet::Overflow),
+#[derive(Debug, thiserror::Error)]
+enum DecoderErrorInt {
+    #[error("Buffer failed to read from source: {0}")]
+    Io(#[from] std::io::Error),
+    #[error("EOF encountered")]
+    Eof,
```

Comment:
Only used internally as the name suggests. The purpose for this is to much easily bubble up a EOF condition to the top-most function `Decoder::next_single` which returns a `DecoderError`.